### DISCUSSION TITLE
Set animation speed (FPS) for Open3D

### DISCUSSION THREAD

**Discussion by timwoelfle** [dfki-ric/pytransform3d#121] [Feature discussions]
Thank you very much for this awesome package, I'm using it to visualize smartphone IMU data and like it a lot! While I've found how to set the animation speed in matplotlib (interval argument, see [here](https://matplotlib.org/stable/api/_as_gen/matplotlib.animation.FuncAnimation.html)), I haven't found any way to set it for Open3D [animate](https://rock-learning.github.io/pytransform3d/_apidoc/pytransform3d.visualizer.Figure.html#pytransform3d.visualizer.Figure.animate). Have I missed something? Otherwise it would be great if this could be integrated so that my animations frames per seconds are on par with the IMU sensor reading frequency!

**Comment by AlexanderFabisch** [comment]
> Thank you very much for this awesome package, I'm using it to visualize smartphone IMU data and like it a lot!

Thanks. :)

> I'm using it to visualize smartphone IMU data and like it a lot! While I've found how to set the animation speed in matplotlib (interval argument, see here), I haven't found any way to set it for Open3D animate. Have I missed something? Otherwise it would be great if this could be integrated so that my animations frames per seconds are on par with the IMU sensor reading frequency!

I think Open3D doesn't directly support this, so this is something that would have to be done in pytransform3d.

As a workaround, I would suggest to modify your animation callback to sleep for the correct amount of time. If you expect that your animation does almost need no time you can just call `import time; time.sleep(seconds as float)`. Otherwise you can make a callable object that keeps track of the time that you have to sleep in between.

**Comment by timwoelfle** [comment]
Thank you very much for your quick reply! Alternatively I could try the whole thing with matplotlib where I can already set FPS with the interval argument, but I'm not sure how to transform a box in a matplotlib animation (I did manage in Open3D)? Unfortunately, this does not work (rather redraws many basis-frames and doesn't transform the box itself during the animation):

```
def update_box(step, n_frames, ax):
    A2B = A2B_from_frame(step)
    plot_transform(ax=ax, A2B=A2B)
    return ax

fig = plt.figure()

ax = make_3d_axis(ax_s=1)

box = plot_box(ax=ax, size=[0.5, 1, 0.05], A2B=A2B, wireframe=True, alpha=0.3, color="k")
plot_transform(ax=ax, A2B=A2B)

anim = animation.FuncAnimation(
       fig, update_box, n_frames, fargs=(n_frames, ax),
       interval=20, blit=False)

plt.show()
```

I've successfully animated a Frame by modifying your [example](https://rock-learning.github.io/pytransform3d/_auto_examples/animations/animate_rotation.html#sphx-glr-auto-examples-animations-animate-rotation-py), but I'm not sure how to add the box:

```
def update_frame(step, n_frames, frame):
    A2B = A2B_from_frame(step)
    frame.set_data(A2B)
    return frame

fig = plt.figure()

ax = make_3d_axis(ax_s=1)

frame = Frame(A2B)
frame.add_frame(ax)

anim = animation.FuncAnimation(
       fig, update_frame, n_frames, fargs=(n_frames, frame),
       interval=20, blit=False)

plt.show()
```

(Sorry, I know this doesn't exactly belong here, but maybe you're able to help :-) )

**Comment by AlexanderFabisch** [comment]
Matplotlib is not the best tool for 3d animations. I'd recommend some real 3d visualization software for more complex cases.

As you can see [here](https://github.com/rock-learning/pytransform3d/blob/master/examples/animations/animate_angle_axis_interpolation.py#L48), you need to modify the Matplotlib (artist?) objects in each frame. These will be created [here](https://github.com/rock-learning/pytransform3d/blob/master/pytransform3d/plot_utils.py#L512). `plot_box` will not return them. I'd recommend to clone the code from `plot_box` if you want to use matplotlib for these animations.

Unforunately, I don't have a good recommendation as an alternative to Matplotlib or Open3D in Python, but you could take a look at pybullet (which is mainly a physics engine, but can also be used for simple real-time rendering). Take a look at the functions
* pybullet.connect(pybullet.GUI) to start the GUI
* pybullet.createVisualShape and pybullet.createMultiBody to create a box
* pybullet.resetBasePositionAndOrientation to set position and orientation (quaternion, scalar last!)
* pybullet.setRealTimeSimulation to render in real time, example: https://github.com/bulletphysics/bullet3/blob/21d1b8fc7183eed5903cf7bfe9ef8a81e432ab15/examples/pybullet/gym/pybullet_examples/biped2d_pybullet.py#L33

**Comment by timwoelfle** [comment]
Thanks a lot, I'll check it out!
